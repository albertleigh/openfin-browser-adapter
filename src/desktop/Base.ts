export abstract class BaseApiClass {

    static staticMethods:string[];
    static staticSyncMethods:string[];
    static instanceMethods:string[];
    static instanceSyncMethods:string[];

    constructor(){

    }
}