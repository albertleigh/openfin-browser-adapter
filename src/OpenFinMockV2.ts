export class OpenFinMockV2{

    static silentMode:boolean = false;

    static generateMethods(name:string,Klass:any){

        const staticMethods = Klass.staticMethods||[];
        const staticSyncMethods = Klass.staticSyncMethods||[];
        const instanceMethods = Klass.instanceMethods||[];
        const instanceSyncMethods = Klass.instanceSyncMethods||[];

        for (const method of staticMethods){
            if (Klass[method]){
                console.warn(`[OpenFinMock] Warning! Attempt to overwrite already implemented static methods:\`${name}.${method}\`. This method will not be overwritten. Remove this method from \`${name}.staticMethods\`.`);
                continue;
            }

            Klass[method] = async (...args:any[])=>{
                let argsMsg = '';

                if (args.length){
                    argsMsg = `This method also received the arguments: ${JSON.stringify(args,null,2)}`;
                }

                if (!OpenFinMockV2.silentMode){
                    console.info(`[OpenFinMock] Static method \`${name}.${method}\` not implemented. 
                        ${argsMsg} This method will not return anything, which will probably have unintended consequence.
                        Implement this method on the \`${name}\` class to return a value.               
                    `);
                }
            }
        }

        for (const method of staticSyncMethods){
            if (Klass[method]){
                console.warn(`[OpenFinMock] Warning! Attempt to overwrite already implemented static async methods:\`${name}.${method}\`. This method will not be overwritten. Remove this method from \`${name}.staticMethods\`.`);
                continue;
            }

            Klass[method] = (...args:any[])=>{
                let argsMsg = '';

                if (args.length){
                    argsMsg = `This method also received the arguments: ${JSON.stringify(args,null,2)}`;
                }

                if (args.length>1 && typeof args[args.length-2]=== 'function'){
                    args[args.length-2]();
                }

                if (!OpenFinMockV2.silentMode){
                    console.info(`[OpenFinMock] Static method \`${name}.${method}\` not implemented. 
                        ${argsMsg} This method will not return anything, which will probably have unintended consequence.
                        Implement this method on the \`${name}\` class to return a value.               
                    `);
                }
            }
        }

        for (const method of instanceMethods){
            if (Klass.prototype[method]){
                console.warn(`[OpenFinMock] Warning! Attemp to overwrite already implemented instance method: \`${name}#${method}\`.
                    This method will not be overwritten. Remove this method from \`${name}.instanceMethods\`.       
                `);
                continue;
            }

            Klass.prototype[method] = async (...args:any[]) => {
                let argsMsg = '';

                if (args.length){
                    argsMsg = `This method also received the arguments: ${JSON.stringify(args,null,2)}`;
                }

                if (!OpenFinMockV2.silentMode){
                    console.info(`[OpenFinMock] Instance method \`${name}.${method}\` not implemented. 
                        ${argsMsg} This method will not return anything, which will probably have unintended consequence.
                        Implement this method on the \`${name}\` class to return a value.               
                    `);
                }

            }
        }

        for (const method of instanceSyncMethods){
            if (Klass.prototype[method]){
                console.warn(`[OpenFinMock] Warning! Attemp to overwrite already implemented instance method: \`${name}#${method}\`.
                    This method will not be overwritten. Remove this method from \`${name}.instanceMethods\`.       
                `);
                continue;
            }

            Klass.prototype[method] = async (...args:any[]) => {
                let argsMsg = '';

                if (args.length){
                    argsMsg = `This method also received the arguments: ${JSON.stringify(args,null,2)}`;
                }

                if (args.length>1 && typeof args[args.length-2]=== 'function'){
                    args[args.length-2]();
                }

                if (!OpenFinMockV2.silentMode){
                    console.info(`[OpenFinMock] Instance method \`${name}.${method}\` not implemented. 
                        ${argsMsg} This method will not return anything, which will probably have unintended consequence.
                        Implement this method on the \`${name}\` class to return a value.               
                    `);
                }

            }
        }

        return Klass;

    }
}