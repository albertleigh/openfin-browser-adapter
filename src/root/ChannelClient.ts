import {BaseApiClass} from './Base';

export class ChannelClient extends BaseApiClass{

    static staticMethods:string[]=[
    ];

    static staticSyncMethods:string[]=[
    ];

    static instanceMethods:string[]=[
        'dispatch',
    ];

    static instanceSyncMethods:string[]=[
        'disconnect',
        'register',
        'remove',
        'setDefaultAction',
        'onError',
        'beforeAction',
        'afterAction',
    ];

}