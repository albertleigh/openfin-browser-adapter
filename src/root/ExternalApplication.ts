import {BaseApiClass} from './Base';

export class ExternalApplication extends BaseApiClass{


    static staticMethods:string[]=[
        'wrap',
    ];

    static staticSyncMethods:string[]=[
        'wrapSync',
    ];

    static instanceMethods :string[]=[
        'addListener',
        'getInfo',
        'on',
        'once',
        'prependListener',
        'prependOnceListener',
        'quit',
        'registerUser',
        'removeAllListeners',
        'removeListener',
    ];

}