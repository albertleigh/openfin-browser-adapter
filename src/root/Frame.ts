import {BaseApiClass} from './Base';

export class Frame extends BaseApiClass{


    static staticMethods:string[]=[
        'getCurrent',
        'wrap',
    ];

    static staticSyncMethods:string[]=[
        'getCurrentSync',
        'wrapSync',
    ];

    static instanceMethods :string[]=[
        'addListener',
        'getInfo',
        'getParentWindow',
        'on',
        'once',
        'prependListener',
        'prependOnceListener',
        'removeAllListeners',
        'removeListener',
    ];

}