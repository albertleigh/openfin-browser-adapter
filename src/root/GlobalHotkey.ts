import {BaseApiClass} from './Base';

export class GlobalHotkey extends BaseApiClass{

    static staticMethods:string[]=[
        'isRegistered',
        'register',
        'unregister',
        'unregisterAll',
    ];

    static instanceMethods :string[]=[
    ];

}