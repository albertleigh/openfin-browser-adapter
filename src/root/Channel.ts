import {BaseApiClass} from './Base';
import {ChannelProvider} from './ChannelProvider';
import {ChannelClient} from './ChannelClient';
import {OpenFinMockV2} from '../OpenFinMockV2'

const _ChannelProvider = OpenFinMockV2.generateMethods('InterApplicationBus.Channel#ChannelProvider',ChannelProvider);
const _ChannelClient = OpenFinMockV2.generateMethods('InterApplicationBus.Channel#ChannelClient',ChannelClient);

export class Channel extends BaseApiClass{

    static staticMethods:string[]=[
        'onChannelConnect',
        'onChannelDisconnect',
    ];

    static instanceMethods:string[]=[];

    static async create(){
        return new _ChannelProvider();
    }

    static async connect(){
        return new _ChannelClient();
    }

}