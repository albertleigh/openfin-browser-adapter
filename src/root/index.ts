import {Application} from './Application'
import {ExternalApplication} from './ExternalApplication'
import {Frame} from './Frame'
import {Notification} from './Notification'
import {Window} from './Window'
import {Clipboard} from './Clipboard'
import {GlobalHotkey} from './GlobalHotkey'
import {InterApplicationBus} from './InterApplicationBus'
import {System} from './System'
import {BaseApiClass} from './Base'

export type BaseApiDict={[key:string]:BaseApiClass};

export const root:BaseApiDict={
    Application:Application,
    ExternalApplication:ExternalApplication,
    Frame:Frame,
    Notification:Notification,
    Window:Window,
    Clipboard:Clipboard,
    GlobalHotkey:GlobalHotkey,
    InterApplicationBus:InterApplicationBus,
    System:System,
};
