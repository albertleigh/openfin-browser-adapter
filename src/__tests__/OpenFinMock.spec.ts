import mockConsole from 'jest-mock-console';

import { BaseApiClass } from '../root/Base';
import { OpenFinMockV2 } from '../OpenFinMockV2';

describe('OpenFinMock',()=>{

    let restoreConsole:Function = null;

    beforeEach(()=>{
        restoreConsole = mockConsole();
    })

    afterEach(()=>{
        restoreConsole()
    })

    it('silentMode off',()=>{

        class SampleClass extends BaseApiClass{

            static staticMethods:string[]=[
                'staticMethod1',
                'staticMethod2',
                'staticMethodImpl',
                'staticMethodCalled',
            ];

            static instanceMethods :string[]=[
                'instanceMethod1',
                'instanceMethod2',
                'instanceMethod3',
                'instanceMethodImpl',
                'instanceMethodCalled',
            ];

            static staticMethodImpl(){

            }

            instanceMethodImpl(){

            }

        }

        const TargetClz = OpenFinMockV2.generateMethods('testClz',SampleClass);

        TargetClz.staticMethodCalled(()=>{},()=>{});

        expect(TargetClz).toMatchSnapshot();
        expect(console.warn).toMatchSnapshot();

        const instance = new TargetClz();

        instance.instanceMethodCalled(()=>{},()=>{});

        expect(console.info).toMatchSnapshot();

    })

})