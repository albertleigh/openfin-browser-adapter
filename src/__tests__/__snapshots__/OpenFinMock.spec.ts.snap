// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenFinMock silentMode off 1`] = `[Function]`;

exports[`OpenFinMock silentMode off 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[OpenFinMock] Warning! Attempt to overwrite already implemented static methods:\`testClz.staticMethodImpl\`. This method will not be overwritten. Remove this method from \`testClz.staticMethods\`.",
    ],
    Array [
      "[OpenFinMock] Warning! Attemp to overwrite already implemented instance method: \`testClz#instanceMethodImpl\`.
                    This method will not be overwritten. Remove this method from \`testClz.instanceMethods\`.       
                ",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`OpenFinMock silentMode off 3`] = `[Function]`;
